### PR TITLE
Added generation GraphQLClient.fsi

### PR DESCRIPTION
**Pull request by anthony-mi**

### CODE REVIEW

File: src/CodeGen.fs [hunk 1]
```diff
@@ -1176,6 +1183,12 @@ type GraphqlErrorResponse = {{ errors: {errorType} list }}
 type {clientName}(url: string, httpClient: HttpClient, options: JsonSerializerOptions) =
     new(url: string, options: JsonSerializerOptions) = {clientName}(url, new HttpClient(), options)
     new(url: string) = {clientName}(url, new HttpClient(), new JsonSerializerOptions())
+    new(url: string)=
+        let options =
+            let options' = JsonSerializerOptions()
+            options'.Converters.Add(JsonFSharpConverter())
+            options'
+        {clientName}(url, new HttpClient(), options)
```
Comment (Zaid-Ajaj) [inline src/CodeGen.fs:1191]: Can you change to the following:
```fs
let options = JsonSerializerOptions()
options.Converters.Add(JsonFSharpConverter())
{clientName}(url, new HttpClient(), options)
```

Comment (anthony-mi) [reply src/CodeGen.fs]: Updated

File: src/CodeGen.fs [hunk 2]
```diff
@@ -1203,4 +1214,73 @@ type {clientName}(url: string, httpClient: HttpClient) =
...
+
+open Newtonsoft.Json
+open Newtonsoft.Json.Linq
+open Fable.Remoting.Json
+open System.Net.Http
+open System.Text
+
+type {clientName} =
+    class
+        /// <summary>Creates {clientName}</summary>
```
Comment (xperiandri) [inline src/CodeGen.fs:1230]:
```suggestion
        /// <summary>Creates {clientName} specifying <see href="T:HttpClient">HttpClient</see> instance</summary>
```

File: src/CodeGen.fs [hunk 3]
```diff
@@ -1203,4 +1214,73 @@ type {clientName}(url: string, httpClient: HttpClient) =
...
+        /// <remarks>
+        /// In order to enable all F# types serialization and deserealization you must add 
+        /// <see href="T:Fable.Remoting.Json.FableJsonConverter">FableJsonConverter</see>
+        /// from <a href="https://github.com/Zaid-Ajaj/Fable.SimpleJson">Fable.SimpleJson</a> NuGet package
```
Comment (xperiandri) [inline src/CodeGen.fs:1234]:
```suggestion
        /// In order to enable all F# types serialization and deserealization 
        /// <see href="T:Fable.Remoting.Json.FableJsonConverter">FableJsonConverter</see> is added
        /// from <a href="https://github.com/Zaid-Ajaj/Fable.SimpleJson">Fable.SimpleJson</a> NuGet package
```

File: src/CodeGen.fs [hunk 4]
```diff
@@ -1203,4 +1214,73 @@ type {clientName}(url: string, httpClient: HttpClient) =
...
+        /// </remarks>
+        /// <param name="url">GraphQL endpoint URL</param>
+        new: url: string * client: HttpClient -> SpotifyGraphqlClient        new: string * HttpClient * JsonSerializerOptions -> SpotifyGraphqlClient
```
Comment (xperiandri) [inline src/CodeGen.fs:1237]:
```suggestion
        new: url: string * client: HttpClient -> SpotifyGraphqlClient
        
        /// <summary>Creates {clientName}</summary>
        /// <remarks>
        /// In order to enable all F# types serialization and deserealization 
        /// <see href="T:Fable.Remoting.Json.FableJsonConverter">FableJsonConverter</see> is added
        /// from <a href="https://github.com/Zaid-Ajaj/Fable.SimpleJson">Fable.SimpleJson</a> NuGet package
        /// </remarks>
        new: url: string -> SpotifyGraphqlClient
```

File: src/CodeGen.fs [hunk 5]
```diff
@@ -1203,4 +1214,73 @@ type {clientName}(url: string, httpClient: HttpClient) =
...
+    $"""namespace {projectName}
+
+open System.Net.Http
+open System.Text
+open System.Text.Json
+open FSharp.Control.Tasks
+
+type {clientName} =
+    class
+        /// <summary>Creates {clientName}</summary>
```
Comment (xperiandri) [inline src/CodeGen.fs:1251]:
```suggestion
        /// <summary>
        /// Creates {clientName} specifying <see href="T:HttpClient">HttpClient</see> instance 
        /// and <see href="T:JsonSerializerOptions">JsonSerializerOptions</see>
        /// </summary>
```

File: src/CodeGen.fs [hunk 6]
```diff
@@ -1176,6 +1183,10 @@ type GraphqlErrorResponse = {{ errors: {errorType} list }}
 type {clientName}(url: string, httpClient: HttpClient, options: JsonSerializerOptions) =
     new(url: string, options: JsonSerializerOptions) = {clientName}(url, new HttpClient(), options)
     new(url: string) = {clientName}(url, new HttpClient(), new JsonSerializerOptions())
+    new(url: string)=
+        let options = JsonSerializerOptions()
+        options.Converters.Add(JsonFSharpConverter())
+        {clientName}(url, new HttpClient(), options)
```
Comment (xperiandri) [inline src/CodeGen.fs:1189]:
```suggestion
    new(url: string, options: JsonSerializerOptions) = {clientName}(url, new HttpClient(), options)
    new(url: string, client: HttpClient) =
        let options = JsonSerializerOptions()
        options.Converters.Add(JsonFSharpConverter())
        {clientName}(url, client, options)
    new(url: string) = {clientName}(url, new HttpClient())
```

File: src/CodeGen.fs [hunk 7]
```diff
@@ -1203,4 +1214,73 @@ type {clientName}(url: string, httpClient: HttpClient) =
...
+type {clientName} =
+    class
+        /// <summary>Creates {clientName}</summary>
+        /// <remarks>
+        /// In order to enable all F# types serialization and deserealization you must add 
+        /// <see href="T:System.Text.Json.Serialization.JsonFSharpConverter">JsonFSharpConverter</see>
+        /// from <a href="https://github.com/Tarmil/FSharp.SystemTextJson">FSharp.SystemTextJson</a> NuGet package
+        /// </remarks>
+        /// <param name="url">GraphQL endpoint URL</param>
+        new: url: string * client: HttpClient * options: JsonSerializerOptions -> SpotifyGraphqlClient        new: string * HttpClient * JsonSerializerOptions -> SpotifyGraphqlClient
```
Comment (xperiandri) [inline src/CodeGen.fs:1258]:
```suggestion
        new: url: string * client: HttpClient * options: JsonSerializerOptions -> SpotifyGraphqlClient
        
        /// <summary>
        /// Creates {clientName} specifying <see href="T:JsonSerializerOptions">JsonSerializerOptions</see>
        /// </summary>
        /// <param name="url">GraphQL endpoint URL</param>
        /// <remarks>
        /// In order to enable all F# types serialization and deserealization <b>you must</b> add 
        /// <see href="T:System.Text.Json.Serialization.JsonFSharpConverter">JsonFSharpConverter</see>
        /// from <a href="https://github.com/Tarmil/FSharp.SystemTextJson">FSharp.SystemTextJson</a> NuGet package yourself
        /// </remarks>
        new: url: string * options: JsonSerializerOptions -> SpotifyGraphqlClient
        
        /// <summary>
        /// Creates {clientName} specifying <see href="T:HttpClient">HttpClient</see> instance
        /// </summary>
        /// <param name="url">GraphQL endpoint URL</param>
        /// <remarks>
        /// In order to enable all F# types serialization and deserealization 
        /// <see href="T:System.Text.Json.Serialization.JsonFSharpConverter">JsonFSharpConverter</see> by default is added
        /// from <a href="https://github.com/Tarmil/FSharp.SystemTextJson">FSharp.SystemTextJson</a> NuGet package
        /// </remarks>
        new: url: string * client: HttpClient -> SpotifyGraphqlClient
        
        /// <summary>Creates {clientName}</summary>
        /// <param name="url">GraphQL endpoint URL</param>
        /// <remarks>
        /// In order to enable all F# types serialization and deserealization 
        /// <see href="T:System.Text.Json.Serialization.JsonFSharpConverter">JsonFSharpConverter</see> by default is added
        /// from <a href="https://github.com/Tarmil/FSharp.SystemTextJson">FSharp.SystemTextJson</a> NuGet package
        /// </remarks>
        new: url: string -> SpotifyGraphqlClient
```

File: src/CodeGen.fs [hunk 8]
```diff
@@ -1203,4 +1214,73 @@ type {clientName}(url: string, httpClient: HttpClient) =
...
+open System.Text.Json
+open FSharp.Control.Tasks
+
+type {clientName} =
+    class
+        /// <summary>Creates {clientName}</summary>
+        /// <remarks>
+        /// In order to enable all F# types serialization and deserealization you must add 
+        /// <see href="T:System.Text.Json.Serialization.JsonFSharpConverter">JsonFSharpConverter</see>
+        /// from <a href="https://github.com/Tarmil/FSharp.SystemTextJson">FSharp.SystemTextJson</a> NuGet package
```
Comment (xperiandri) [inline src/CodeGen.fs:1255]:
```suggestion
        /// In order to enable all F# types serialization and deserealization <b>you must</b> add 
        /// <see href="T:System.Text.Json.Serialization.JsonFSharpConverter">JsonFSharpConverter</see>
        /// from <a href="https://github.com/Tarmil/FSharp.SystemTextJson">FSharp.SystemTextJson</a> NuGet package yourself
```

File: src/CodeGen.fs [hunk 9]
```diff
@@ -1203,4 +1214,73 @@ type {clientName}(url: string, httpClient: HttpClient) =
...
+    class
+        /// <summary>Creates {clientName}</summary>
+        /// <remarks>
+        /// In order to enable all F# types serialization and deserealization you must add 
+        /// <see href="T:Fable.Remoting.Json.FableJsonConverter">FableJsonConverter</see>
+        /// from <a href="https://github.com/Zaid-Ajaj/Fable.SimpleJson">Fable.SimpleJson</a> NuGet package
+        /// </remarks>
+        /// <param name="url">GraphQL endpoint URL</param>
+        new: url: string * client: HttpClient * options: JsonSerializerOptions -> SpotifyGraphqlClient        new: string * HttpClient * JsonSerializerOptions -> SpotifyGraphqlClient
+    end
```
Comment (xperiandri) [inline src/CodeGen.fs:1280]: Analogical to previous

**Review by xperiandri — CHANGES_REQUESTED** [review]

File: src/CodeGen.fs (hunk 3 above)
Comment (Zaid-Ajaj) [reply src/CodeGen.fs]: These remarks are not entirely correct, the type `FableJsonConverter` from `Fable.Remoting.Json` is not from `Fable.SimpleJson`. 
 - Fable.Remoting.Json is a F# JSON converter running on dotnet
 - Fable.SimpleJson is F# JSON converter running with Fable

Comment (anthony-mi) [reply src/CodeGen.fs]: Ok, I'll redo

File: src/Program.fs [hunk 10]
```diff
@@ -456,14 +457,17 @@ let generate (configFile: string) =
             match config.target with
             | OutputTarget.Fable ->
                 let graphqlClientPath = Path.GetFullPath(Path.Combine(config.output, fileName "GraphqlClient.fs"))
+                let graphqlClientFsiPath = Path.GetFullPath(Path.Combine(config.output, fileName "GraphqlClient.fsi"))
                 colorprintfn "✏️  Generating GraphQL client $green[%s]" graphqlClientPath
-                generatedFiles.Add(graphqlClientPath)
+                generatedFiles.AddRange([| graphqlClientPath; graphqlClientFsiPath|])
```
Comment (Zaid-Ajaj) [inline src/Program.fs:462]: the fsi file must be compiled before the the actual implementation file:
```fs
generatedFiles.AddRange [|
    graphqlClientFsiPath
    graphqlClientPath
|]
```

File: src/Program.fs [hunk 11]
```diff
@@ -599,10 +609,16 @@ let generate (configFile: string) =
...
+                    then 
+                        seq {
+                            MSBuildXElement.Compile($".\{config.project}.GraphqlClient.fs")
+                            MSBuildXElement.Compile($".\{config.project}.GraphqlClient.fsi")
+                        }
+                    else
+                        seq {
+                            MSBuildXElement.Compile($".\{outputDirectoryName}\\dotnet\{config.project}.GraphqlClient.fs")
+                            MSBuildXElement.Compile($".\{outputDirectoryName}\\dotnet\{config.project}.GraphqlClient.fsi")
+                        }
```
Comment (Zaid-Ajaj) [inline src/Program.fs:621]: Same as above, the `fsi` definition comes before the implementation

File: src/CodeGen.fs [hunk 12]
```diff
@@ -1203,4 +1214,114 @@ type {clientName}(url: string, httpClient: HttpClient) =
 let sampleFSharpGraphqlClient projectName clientName errorType members serializer =
     match serializer with
     | SerializerType.System -> sampleFSharpSystemGraphqlClient projectName clientName errorType members
-    | SerializerType.Newtonsoft -> sampleFSharpNewtonsoftGraphqlClient projectName clientName errorType members
\ No newline at end of file
+    | SerializerType.Newtonsoft -> sampleFSharpNewtonsoftGraphqlClient projectName clientName errorType members
+
+let sampleFableGraphqlClientFsi projectName clientName =
```
Comment (Zaid-Ajaj) [inline src/CodeGen.fs:1219]: This function is implemented incorrectly, Fable fsi implementation doesn't use any of `Newtonsoft.Json`, `Fable.Remoting.Json`, `System.Net.Http` and definitely no `HttpClient`

Comment (anthony-mi) [reply src/CodeGen.fs:1219]: @Zaid-Ajaj updated

**Comment by xperiandri** [comment]
@Zaid-Ajaj can I use FusionTasks NuGet to simplify code?

`task` computation expression does not require `Async.AwaitTask`, `async` computation expression requires it.
With FusionTasks we can use the same code and simplify generation

**Comment by Zaid-Ajaj** [comment]
@xperiandri task is already included when asyncReturnType: "task" in which case the Ply package is included

**Comment by xperiandri** [comment]
I need https://www.nuget.org/packages/FSharp.Control.FusionTasks/2.2.0

**Comment by Zaid-Ajaj** [comment]
You mean you want to use FusionTasks to simplify the async implementation? I don't really think adding a dependency just for Async.AwaitTask is worth it.

If you _need_ FusionTasks in your application, why not simply add it in your project that references the generated Snowflaqe client?

**Comment by xperiandri** [comment]
It worse because now we have 2 generator functions everywhere async/task appear which complecates the code a lot and adds more chances for mistake

**Comment by Zaid-Ajaj** [comment]
I agree it is a bit complicated but it makes the result nicer for users who want Task<'T> as the output of the client function, not just for simplifying the implementation details. 

I think implementing error handling option from #29 will be the last hurdle and it's still doable IMO

Do you have any other features might be impacted because of the task/option?

**Comment by xperiandri** [comment]
I don't know why could people not use https://www.nuget.org/packages/FSharp.Control.FusionTasks/2.2.0
It is one of the first NuGet packages I install to any F# project. Adding Async.AwaitTask to every async .NET BCL call is crazy...

**Comment by xperiandri** [comment]
The same as manually calling `OnPropertyChanged` instead of installing `PropertyChanged.Fody` and make it done automatically

**Comment by Zaid-Ajaj** [comment]
> I don't know why could people not use https://www.nuget.org/packages/FSharp.Control.FusionTasks/2.2.0
It is one of the first NuGet packages I install to any F# project. Adding Async.AwaitTask to every async .NET BCL call is crazy...

Users of Snowflaqe don't have to call `Async.AwaitTask`, it is just the generated code. You can always install FusionTasks into your own projects whenever you need it. 

Sticking with `Ply` for now as a dependency of the generated projects because it gives a nice `task` CE to work with when tasks are preferred.

**Comment by Numpsy** [comment]
> 
> Sticking with `Ply` for now as a dependency of the generated projects because it gives a nice `task` CE to work with when tasks are preferred.

One of the reasons I made the task change was that 'fits' more when the actual 'async' functionality is using tasks anyway (i.e. HttpClient and the Json libs are task based), but then I was also having a go at inserting the generated lib into an existing C# stack and the tasks returned from Ply just work directly there which is nice for interop.

(I'm a massive beginner with F# and had never heard of FusionTasks, but it looks to me like ```task { }``` is going to end up part of FSharp Core at some point, at which point the issue and the requirement for 3rd party libs might just go away? (or maybe i'm being hopeful looking at how long that question has been running for!))

**Comment by Zaid-Ajaj** [comment]
Hi @anthony-mi what's the status of this PR? Any plans to continue working on?

@xperiandri if there are issues or discussions outside of this PR, please consider opening a separate issue to discuss it there

**Comment by xperiandri** [comment]
@Zaid-Ajaj Anthony decided to exit from the project. The other developer will pick up the work in several weeks.

**Comment by Zaid-Ajaj** [comment]
Thanks for the update @xperiandri does this mean this PR can closed to start all over later on or should I keep it open for now?

**Comment by xperiandri** [comment]
Keep it open for now. I will let you know

**Comment by xperiandri** [comment]
@Zaid-Ajaj we have discovered that in order to have FSI we need to declare all public members within it too

**Comment by xperiandri** [comment]
So that we think of removing FSI and rewriting the client class without the default constructor

**Comment by Zaid-Ajaj** [comment]
> So that we think of removing FSI and rewriting the client class without the default constructor

Sounds like a good idea for now. Proper FSI support can be added in another PR

**Comment by xperiandri** [comment]
We will have a look closer soon. Removing FSI looks easier to implement an maintain for now

**Pull request merged**